drivers: clock_control: Introduce revision 3 of CCM driver

LaurentiuM1234 opened the pull request:
This commit introduces a new version of the CCM driver. The purpose of this is to make the CCM driver usable by all NXP SoCs despite the changes in the HAL API.

With this driver, each SoC will have to define its own clock tree and a few common operations such as clock enable/disable, set_rate and get_rate.

Hopefully, in time, this driver will replace clock_control_mcux_ccm.c and clock_control_mcux_ccm_rev2.c.

danieldegrasse reviewed (commented):
Thanks for this contribution- I left a few initial comments after review.

Key questions I have:
- does this driver intend to enable the user to configure the PLL frequencies at runtime?
- if we implement the `set_rate` functionality, but do not allow reconfiguring the MUX selection or PLL frequencies, will we lock the range of frequencies `set_rate` supports given the frequency of the PLL and the mux that was configured at boot time?

LaurentiuM1234 commented on include/zephyr/drivers/clock_control/clock_control_mcux_ccm_rev3.h:
```diff
@@ -0,0 +1,149 @@
+	mm_reg_t regmap;
+	mm_reg_t pll_regmap;
+
+	uint32_t ipc_handle;
```
This will be used in the case of i.MX8QM where the CCM is configured through SCFW. If the number of fields in `struct imx_ccm_data` and `struct imx_ccm_config` gets out of hand we can always add a `void *pdata` field which will hold the SoC-specific data.

LaurentiuM1234 commented on include/zephyr/drivers/clock_control/clock_control_mcux_ccm_rev3.h:
```diff
@@ -0,0 +1,149 @@
+ * if successful, the function will return the new rate which
+ * may differ from the requested rate.
+ */
+int imx_ccm_clock_set_rate(const struct device *dev, struct imx_ccm_clock *clk, uint32_t rate);
```
Exactly. Currently, I'm working on making i.MX93 switch over to CCM rev3. I was going to wait until this PR got merged but I guess it would be a good idea to submit it earlier so people could see an example of how these functions can be implemented.

danieldegrasse replied:
I think submitting it as a commit within this PR makes sense. I find evaluating the entire implementation difficult otherwise

LaurentiuM1234 commented on include/zephyr/drivers/clock_control/clock_control_mcux_ccm_rev3.h:
```diff
@@ -0,0 +1,149 @@
+};
+
+struct imx_ccm_fixed {
+	char *name;
```
This is primarily used in the clock tree description for readability. It might be easier to understand it if the clock is also assigned a string name and not just an ID from the NXP HAL.

danieldegrasse replied:
I'm wondering if we can use macros defined in a header file for the clock ID values. This way we avoid the overhead of the name constants being stored, but the clock tree defined in the dts file is still using more friendly clock names

LaurentiuM1234 commented on include/zephyr/drivers/clock_control/clock_control_mcux_ccm_rev3.h:
```diff
@@ -0,0 +1,149 @@
+struct imx_ccm_pll {
+	char *name;
+	uint32_t id;
+	uint32_t max_freq; /* maximum allowed frequency */
```
Hm, configuring the PLLs is somewhat delicate IMO as if it's done during runtime we need to take into account the fact that other subsystems may be affected. I'd say it would be easier to configure them during driver initialization (possibly with the help of some devicetree properties). Either way, the way they are configured should be entirely up to the SoC developer. There are no restrictions ATM regarding this.

LaurentiuM1234 commented:
@danieldegrasse Thanks for the feedback.




> * does this driver intend to enable the user to configure the PLL frequencies at runtime?

Please see answer inline.


> * if we implement the `set_rate` functionality, but do not allow reconfiguring the MUX selection or PLL frequencies, will we lock the range of frequencies `set_rate` supports given the frequency of the PLL and the mux that was configured at boot time?

If you don't want to change the MUX and the PLL configuration then yes you're stuck with the range of frequencies resulting from only changing DIV. Whether you want to do this or not should be entirely up to you.

LaurentiuM1234 commented:
Updates:

- added an example of how this driver could be used on i.MX93.
- changed PR to draft since now there are some HAL dependencies which I'll take care of if people decide this change can be upstreamed.

LaurentiuM1234 commented:
FYI: another example of how the CCM Rev3 driver may be used can be seen in https://github.com/zephyrproject-rtos/zephyr/pull/60136

dbaluta commented on drivers/clock_control/CMakeLists.txt:
```diff
@@ -11,6 +11,7 @@ zephyr_library_sources_ifdef(CONFIG_CLOCK_CONTROL_LPC11U6X            clock_cont
 zephyr_library_sources_ifdef(CONFIG_CLOCK_CONTROL_MCHP_XEC            clock_control_mchp_xec.c)
```
@LaurentiuM1234 would also like to see in the commit message what CCM v3 driver brings in over v2.

LaurentiuM1234 replied:
Sure, can do so. The main advantages of CMM v3 over the pervious versions are:

1. Scalability + readability. Since pervious versions heavily rely on macros, adding support for it on more SoCs will heavily impact its readability as you'd need to use even more macros. This is because the HAL API may change from SoC to SoC.
2. Flexibility. SoC developers may choose to implement the clock control operations for their SoC however they see fit. 
3. Ability to set a clock's frequency.

dbaluta commented on drivers/clock_control/clock_control_mcux_ccm_rev3.c:
```diff
@@ -0,0 +1,102 @@
+
+	clock_name = (uintptr_t)sys;
+	cfg = dev->config;
+
```
Is this really `clock_name` here. It sounds to me that is more like `clk_num`.

LaurentiuM1234 replied:
It's called `clock_name` to match the name of the clock cell. However, we can change this to `clk_num` if you think it will add more readability to the driver.

LaurentiuM1234 committed:
modules: Kconfig.imx: Add configuration for disabling implicit clocking

By default, the NXP HAL functions may also perform clock management
operations such as gating or ungating clocks. Since this behaviour
is not always desired, this commit introduces a configuration meant
to allow users to disable it.

To make sure existing applications are not broken, this behaviour
is enabled by default.

Signed-off-by: Laurentiu Mihalcea <laurentiu.mihalcea@nxp.com>

LaurentiuM1234 committed:
drivers: clock_control: Add revision 3 of CCM driver

This commit introduces revision 3 of the CCM driver. The purpose
of this new driver is to provide an interface to the clock controller
API that will remain unchanged despite the used variants of the NXP HAL
clock drivers. This new interface shall also make implementing
the CCM driver for a new i.MX SoC much easier as most of the complexity
is kept in the interface instead of the SoC layer.

Signed-off-by: Laurentiu Mihalcea <laurentiu.mihalcea@nxp.com>

LaurentiuM1234 committed:
nxp: mimx93: Switch to using CCM Rev3

With this commit, the i.MX93 SoC will start using CCM Rev3. To make
this work, the following changes had to be made:
	1) Removed CCM-related nodes from the DTS.
		* no longer needed.
	2) Removed static memory mapping using mmu_regions.c.
		* no longer needed, CCM Rev3 uses device_map().
	3) Implemented the CCM Rev3 basic operations.
	4) Added clock ungating operation to the LPUART driver.
		* this is needed because clocks are gated by
		default. As such, before configuring the LPUART
		module, the clock has to be ungated.

Signed-off-by: Laurentiu Mihalcea <laurentiu.mihalcea@nxp.com>

LaurentiuM1234 committed:
manifest: Update hal_nxp

Update hal_nxp to contain the CCM Rev3-related
changes for i.MX93. This is only temporary. When
the HAL changes get merged, the manifest will
point to a hash instead of a PR.

Signed-off-by: Laurentiu Mihalcea <laurentiu.mihalcea@nxp.com>

LaurentiuM1234 commented:
**V2 changes**:

- driver was restructured completely. ATM, it offers the following features:
    1) Support for PLL initialization.
    2) Linux-like clock initialization and parent assignment using the `asssign-clock*` properties which can be scattered through the DTS.
    3) Support for clocks which are already initialized by another instance (e.g: the ROM code for i.MX93)
    4) Support for ungating clocks for which there's no Zephyr native drivers.

- for details regarding the interface layer please see `clock_control_mcux_ccm_rev3.h`. For SoC layer details and tutorial-like information please see `imx93_ccm.c`.
- for clock configuration examples please see `imx93_ccm.c`.

Please let me know if there's any pieces of documentation that are unclear and/or could be improved.

danieldegrasse commented on boards/arm64/mimx93_evk/mimx93_evk_a55.dts:
```diff
@@ -36,4 +36,7 @@
 	pinctrl-0 = <&uart2_default>;
 	pinctrl-names = "default";
+	assigned-clocks = <IMX93_CCM_LPUART2_ROOT>;
+	assigned-clock-parents = <IMX93_CCM_OSC_24M>;
```
The deviation from the `clocks = <xx>` standard property here concerns me. I think we should still be using the `clocks` property for clock data, unless there is a good reason to deviate from the standard property nameused by other SOCs

LaurentiuM1234 replied:
The problem with the `clocks` property is that it doesn't offer enough flexibility. For instance, the `assigned-clock-parents` property is optional (unless a node needs this property in which case it becomes mandatory). If we were to use the `clocks` then it would be mandatory to specify a clock's parent as the number of elements in the resulting clocks array would have to be divisible by some value. Also, the drivers don't need to have access to all of the fields (e.g: the clock's frequency or the clock's parent). A handle to their clock is the only thing they need.

danieldegrasse commented on boards/arm64/mimx93_evk/mimx93_evk_a55_sof.dts:
```diff
@@ -85,4 +85,7 @@
 	pinctrl-0 = <&uart2_default>;
 	pinctrl-names = "default";
+	assigned-clocks = <IMX93_CCM_LPUART2_ROOT>;
+	assigned-clock-parents = <IMX93_CCM_OSC_24M>;
```
It seems like what we are trying to do here is encode a clock tree with these two properties. Rather than having a `clock` and `clock-parent` property, why not describe the `clock` node as a devicetree node, then have the parent of that node be the source clock?

dbaluta replied:
> It seems like what we are trying to do here is encode a clock tree with these two properties. Rather than having a `clock` and `clock-parent` property, why not describe the `clock` node as a devicetree node, then have the parent of that node be the source clock?

I think we are trying to follow here the Linux implementation which is proved to work for a while.

danieldegrasse replied:
Ok, I've taken a look at the Linux implementation, and I think I understand the goals here a bit better now. With that being said, Linux clearly has different clock drivers for each SOC (since the clock driver itself encodes the clock tree data, and as far as I can see Linux provides a generic framework to handle a lot of common functionalities with clocks: https://docs.kernel.org/driver-api/clk.html.)

Given that, wouldn't we want this PR to have two components:
- a common clock framework of some form (that we can initially use with NXP clock drivers, then attempt to expand to be a generic functionality)
- an SOC specific driver for the iMX 93 part

danieldegrasse replied:
Additionally- just a general question about Linux clock setup- how does Linux handle a case where the initial clock frequency of the SOC should be produced via a PLL (something like 24MHz OSC -> PLL0 -> output clock)? Is the idea that the clock framework would call `set_rate` for the PLL, and then the PLL divider and multiplier would be calculated at runtime?

My concern with this approach in general is that for larger SOCs like the iMX RT series (or iMX itself), this approach makes sense for the ease of use. However, on smaller cores like the LPC parts I am concerned about the overhead of performing these clock calculations on boot

LaurentiuM1234 replied:
> Ok, I've taken a look at the Linux implementation, and I think I understand the goals here a bit better now. With that being said, Linux clearly has different clock drivers for each SOC (since the clock driver itself encodes the clock tree data, and as far as I can see Linux provides a generic framework to handle a lot of common functionalities with clocks: https://docs.kernel.org/driver-api/clk.html.)
> 
> Given that, wouldn't we want this PR to have two components:
> 
> * a common clock framework of some form (that we can initially use with NXP clock drivers, then attempt to expand to be a generic functionality)
> * an SOC specific driver for the iMX 93 part

It does. `clock_control_mcux_ccm_rev3.c` is supposed to contain the SoC-independent operations such as parsing the clock tree to set a clock's rate, managing the software states of the clocks (`GATED`, `UNGATED`) etc....(basically the common clock framework). The SoC specific driver is in `imx93_ccm.c` which implements the SoC-specific operations such as setting a clock's rate, setting a clock's parent etc...


> Additionally- just a general question about Linux clock setup- how does Linux handle a case where the initial clock frequency of the SOC should be produced via a PLL (something like 24MHz OSC -> PLL0 -> output clock)? Is the idea that the clock framework would call `set_rate` for the PLL, and then the PLL divider and multiplier would be calculated at runtime?

Not sure how helpful I can be here since I'm not entirely familiar with how Linux initializes the clocks (also this approach isn't meant to be a 1:1 recreation of the Linux CCF, but rather a very simplified version that would still be powerful enough to serve our purposes). Not sure if this would help but the way I see things there's 3 possible approaches:

1. During initialization every clock is set to a default configuration which can then be altered later on using the `assigned-clock*` properties.
2. During initialization only clocks configured through `assigned-clock*` properties are made usable (also the clocks which would be configured through propagation). This would leave unconfigured clocks basically unusable so the driver should raise an error when attempting to use a clock which depends on one of these clocks.
3. The SoC layer itself knows the default configuration of the clocks so it just informs the upper layer such that no initialization is required.

I tend to think option 1 (or even 3 maybe?) would be more plausible in Linux's case. As a side note, my current approach is based on number 2. Ideally, it should be based on 1 or 3.

> My concern with this approach in general is that for larger SOCs like the iMX RT series (or iMX itself), this approach makes sense for the ease of use. However, on smaller cores like the LPC parts I am concerned about the overhead of performing these clock calculations on boot

Yeah this is indeed concerning. We could optimize the driver such that it initializes only clocks used by consumers, instead of initializing all of the clocks declared in the SoC layer's clock tree. Other than that, you could reduce the overhead by using a toned-down version of the clock tree and reducing the amount of configurations that need to be done through `assigned-clock*` properties.

danieldegrasse replied:
> It does. clock_control_mcux_ccm_rev3.c is supposed to contain the SoC-independent operations such as parsing the clock tree to set a clock's rate, managing the software states of the clocks (GATED, UNGATED) etc....(basically the common clock framework). The SoC specific driver is in imx93_ccm.c which implements the SoC-specific operations such as setting a clock's rate, setting a clock's parent etc...

This makes sense- thank you for the clarification

> 1. During initialization every clock is set to a default configuration which can then be altered later on using the assigned-clock* properties.
2. During initialization only clocks configured through assigned-clock* properties are made usable (also the clocks which would be configured through propagation). This would leave unconfigured clocks basically unusable so the driver should raise an error when attempting to use a clock which depends on one of these clocks.
3. The SoC layer itself knows the default configuration of the clocks so it just informs the upper layer such that no initialization is required.

Of these, I actually prefer 2), since that would result in the lowest runtime overhead for the platform, and the least amount of definitions needed in code.

The key thing I am worried about here is a need to implement `set_rate` on all clock devices. For example, if we define the clock tree itself in the DTS, we could end up with something like this for a PLL:

```
osc_24m {
	clock-frequency = <24000000>;
};

rtc_32k {
	clock-frequency = <32768>;
};


pll_clk_sel {
	input-sources = <&osc_24m &rtc_32k>;
	pll0: pll0 {
		mult = <25>;
		div = <1>;
	};
};
```

with something like this, the SOC initialization code could easily determine the PLL settings needed. If we used `set_rate`, then the SOC code would need to calculate the PLL multiplier and divider at runtime. The other issue I am concerned about is code size. With a devicetree, the C code will come down to a function call (and could be as simple as a few register writes on some platforms). With `set_rate` implementations, the code will need to include logic to perform these calculations at runtime

danieldegrasse commented on drivers/clock_control/clock_control_mcux_ccm_rev3.c:
```diff
@@ -0,0 +1,555 @@
+	 */
+	if (clk->freq == rate) {
+		LOG_ERR("clock %s already set to rate %u", clk->name, rate);
+		return -EALREADY;
```
Why return an error here? The clock is at the rate we need, so I would argue the function can just `return 0` here, since the clock is configured how the user needs it to be

LaurentiuM1234 replied:
According to https://docs.zephyrproject.org/latest/hardware/peripherals/clock_control.html, set_rate() should return -EALREADY if clock was already configured to requested frequency. Despite this, I agree with you but we should try to abide by the clock controller API specification.

danieldegrasse commented on drivers/clock_control/clock_control_mcux_ccm_rev3.c:
```diff
@@ -0,0 +1,555 @@
+	}
+
+	/* go up the clock hierarcy in order to set the parent's rate */
+	ret = _mcux_ccm_set_rate(dev, clk->parent, parent_rate);
```
I'm unclear why we go to the clock parent to set the rate- will this cause dependency problems if the parent clock has other child clock dependencies?

LaurentiuM1234 commented on include/zephyr/drivers/clock_control/clock_control_mcux_ccm_rev3.h:
```diff
@@ -0,0 +1,467 @@
+ *				> this function gates only the clock
+ *				passed as argument. The reason for this
+ *				behaviour is because, generally clocks
+ *				from higher levels are depended upon by
```
Thanks for pointing this out! It was mean to say: `generally clocks from higher levels are depended upon by multiple clocks from lower levels`. Will fix in the next iteration.

dbaluta commented:
@mfischer Hi, Care to have a look? I remeber you had a talk about clocks at ZDS.

danieldegrasse reviewed (commented):
I think rather than sprinkling comments throughout the PR, it would make more sense to highlight the specific concerns I have with this approach to making clocking more generic

If we are going to implement clocking using runtime calculations like this, I think we should probably implement more of the ideas present in the common clock framework. In particular, I think we should probably implement support for things like notifications when clock rates change- without features like this, I don't think we can safely change most clock parent assignments. That will prevent utilization of a lot of low power functionality on SOCs when using this framework.

However, my primary concern is that by implementing more of the common clock framework we will end up with a rather large subsystem that isn't usable by smaller SOCs, which may need to reconfigure clock sources for features like DVFS.

The core use cases I am aware of for expanding the clock framework are the following:
1. support better power management (features like DVFS)
2. enable dynamically setting clock rates for peripherals
Please feel free to highlight other use cases here, if you are aware of them

From my point of view, we can support 1) relatively well with the following rule: if a peripheral is the only consumer of its parent divider/mux, we can freely change that mux/divider setting. Otherwise, we don't permit changing the frequency via the clock_control API.

Use case 2) is harder to support, but I think the idea of discrete set points (where each set point corresponds to a clock tree configuration) makes sense. I've opened a draft PR on the LPC55S69 (https://github.com/zephyrproject-rtos/zephyr/pull/62946) to demonstrate how clocks could be configured using a devicetree based clock tree. We could logically extend this to support multiple clock tree "set points" which reconfigure the clock tree to enable better power savings at reduced frequencies.

LaurentiuM1234 commented on include/zephyr/drivers/clock_control/clock_control_mcux_ccm_rev3.h:
```diff
@@ -0,0 +1,467 @@
+ *				be ignored. Return code is 0.
+ *
+ *			II) mcux_ccm_on()
+ *				> this function ungates all relatives
```
The wording is just so bad, will have to redo the documentation. The term `relatives` is used to refer to the clock's parent, grand-parent etc.... For example, assuming the following tree:

```
A ---> B ---> C
```

where PARENT(C) = B, PARENT(B) = A. Trying to do a `UNGATE(C)` will result in also ungating B and A.

LaurentiuM1234 commented on include/zephyr/drivers/clock_control/clock_control_mcux_ccm_rev3.h:
```diff
@@ -0,0 +1,467 @@
+ *              setting D's rate will result in attempting
+ *              to also set B and A's rates in the following
+ *              order:
+ *			1) set_rate(A)
```
The rate setting starts from the most distant parent which in your example would theoretically be OSC. Also, you don't propagate the child's rate, but rather the parent's theoretical rate (i.e: the rate required for the parent such that the child has the requested rate).

As things are now, according to the SOC code, the `set_rate` operation would stop at your divider (a.k.a the root clock) (that is if you start the operation from the IP clock).

LaurentiuM1234 commented on include/zephyr/drivers/clock_control/clock_control_mcux_ccm_rev3.h:
```diff
@@ -0,0 +1,467 @@
+ *	a) Computing a clock's level in the clock tree
+ *		- a clock's level is computed as the
+ *		number of parents one needs to traverse
+ *		to reach the NULL parent.
```
The clock that has a NULL parent is a root clock. For example:

```
A ---> B ---> C
```

where PARENT(C) = B and PARENT(B) = A, A is the root clock and has PARENT(A) = NULL. There's no point in encoding this information through another variable.

LaurentiuM1234 commented:
> The core use cases I am aware of for expanding the clock framework are the following:
> 
> 1. support better power management (features like DVFS)
> 2. enable dynamically setting clock rates for peripherals
>    Please feel free to highlight other use cases here, if you are aware of them

Mm, as far as I can tell, dynamically changing clock rates is a nice to have feature but the cases in which you'd actually need to do this may not be many, so I wouldn't say this is the main requirement for the clock framework.

From my POV, the clock framework should aim to do the following things:

1. Try to reduce duplicate code as much as possible.
2. Try to make the lives of the developers as easy as possible by implementing most of the complexity in the upper layers and leaving only the basic SoC-specific operations to the SoC layer.
3. Try to make the lives of the users as easy as possible. They shouldn't have to care about the entire clock hierarchy. All they need to care about is configuring the clocks used by their IPs (if the default configuration is not ok)
4. Try to be as flexible as possible to allow multiple use-cases. For example, in one case you might need `FREQ(IP_CLOCK_X) = Y`, while in other cases you may need `FREQ(IP_CLOCK_X) = y`

IMO, the current implementation is pretty mediocre when talking about points 2 and 3 (possibly even 1) so until I address this the patch series is not in a mergeable state.

> I've opened a draft PR on the LPC55S69 (#62946) to demonstrate how clocks could be configured using a devicetree based clock tree. We could logically extend this to support multiple clock tree "set points" which reconfigure the clock tree to enable better power savings at reduced frequencies.

Thanks for sharing this with me! I'm not sure I fully understand the idea of `set points`. Do you mean something like a clock tree snapshot or configuration? For example:

```
A ---> B ---> C
```

would be the first configuration, while

```
D ---> B ---> C
```

would be the second configuration supported? (assuming B can have A and D as parents). Would you have a `clock_init()` function for each of these configurations? At a first glance, the problem I see with this approach is that we might end up with a lot of files like `lpc55Sxx_clocks.c` (1 for each SoC or maybe clock tree?), which seems like a pain to write it yourself (unless you somehow manage to do some macro magic which in itself is a pain to do) (maybe generate it?). Also, for example, the i.MX93 CCM has a bunch of clocks, won't that make the `clock_init()` function basically unreadable if we were to encode the whole tree in the DTS?

mfischer commented:
> @mfischer Hi, Care to have a look? I remeber you had a talk about clocks at ZDS.

I'll hopefully get around to it. $dayjob is crazy at the moment and I was out of office the last two weeks. Appreciate the tag.

danieldegrasse commented:
> Thanks for sharing this with me! I'm not sure I fully understand the idea of set points. Do you mean something like a clock tree snapshot or configuration? For example:

Yes, exactly- each "set point" would be a snapshot of clock settings. For example, set point 1 might configure the core system clock to use PLL0, and set point 2 might migrate the system clock to a low frequency internal oscillator (for low power consumption)

> Would you have a clock_init() function for each of these configurations? At a first glance, the problem I see with this approach is that we might end up with a lot of files like lpc55Sxx_clocks.c (1 for each SoC or maybe clock tree?)

Yes, we would absolutely have a file like `lpc55Sxx_clocks.c` for most SoCs. We would not need one for each clock tree, the idea would be that we can initialize some form of `struct` describing the clock state for a given clock tree (each `struct` inst would be considered a "set point"), and then switch between the "set points" at runtime.

>  which seems like a pain to write it yourself (unless you somehow manage to do some macro magic which in itself is a pain to do) (maybe generate it?)

This is a concern I have as well. The clock tree itself is automatically generated, and I may look at how to autogenerate the clock initialization code as well (since it really is just configuring dividers and muxes for the most part, the only bits that need much hand editing are the PLL setup). The benefit of this approach is that after optimization, the `clock_init` function can come down to 2-3 function calls to setup the core clock source and switch the core clock mux to that option.

Truthfully, I think your approach is a lot more developer-friendly. Looking at the CCF in Linux, it seems like code reuse is very high (IE you can reuse the clock driver for a PLL if the IP is reused). Part of the issue I expect us to run up against with code reuse here is that our `fsl_clock.h` APIs have a *nasty* habit of changing between SOC families, even if the underlying clock structure is similar enough

danieldegrasse commented:
> At a first glance, the problem I see with this approach is that we might end up with a lot of files like lpc55Sxx_clocks.c (1 for each SoC or maybe clock tree?), which seems like a pain to write it yourself (unless you somehow manage to do some macro magic which in itself is a pain to do) (maybe generate it?). Also, for example, the i.MX93 CCM has a bunch of clocks, won't that make the clock_init() function basically unreadable if we were to encode the whole tree in the DTS?

Yes, you definitely will- however, I'm not sure how we avoid that in either case. With something like this design, we still need some type of clock tree definition in C data structures, right? The reason I took the setpoint approach is to optimize flash usage as much as possible. One way we can do this is to avoid pulling in any clock configuration code from the `fsl_clock.c` driver that we don't need in order to apply clock settings requested by the user

I've opened another PR in pursuit of this, #66732. This no longer uses the "hierarchical" clock structure in devicetree, but instead uses a compatible file to define all possible clock setpoint attributes. This file is being generated from the MCUXpresso configuration tools clock data (as was the devicetree clock structure). This PR demonstrates how setpoints could be supported. Each setpoint is implemented by a function, which is defined from an (admittedly very ugly) macro: https://github.com/zephyrproject-rtos/zephyr/pull/66732/files#diff-3e9e59adc512d595d49ada8ed3c5e92a0d6702a5cd0e924783ae1e483b0ba5cbR211. The advantage of this approach is flash savings: one clock function configuring a set point takes around 200 bytes of flash, and there is no need to link in clock configuration functions that won't be used (which we would have to do if we parsed the clock configuration structures at runtime)
